OptimizelySDK/Event/BatchEventProcessor.cs

```diff
@@ -0,0 +1,358 @@
+            
+            Executer = new Thread(() => Run());
+            Executer.Start();
+            IsStarted = true;
+        }
+
+        /// <summary>
+        /// Scheduler method that periodically runs on provided
+        /// polling interval.
+        /// </summary>
+        public virtual void Run()
+        {
+            try
+            {
+                while (true)
+                {
+                    if (DateTime.Now.Millisecond > FlushingIntervalDeadline)
+                    {
+                        Logger.Log(LogLevel.DEBUG, "Deadline exceeded flushing current batch.");
+                        FlushQueue();
+                    }
+                    
+                    if (!EventQueue.TryTake(out object item, 50))
+                    {
+                        Logger.Log(LogLevel.DEBUG, "Empty item, sleeping for 50ms.");
+                        Thread.Sleep(50);
+                        continue;
+                    }
+
+                    if (item == SHUTDOWN_SIGNAL)
+                    {
+                        Logger.Log(LogLevel.INFO, "Received shutdown signal.");
+                        break;
+                    }
+
+                    if (item == FLUSH_SIGNAL)
+                    {
+                        Logger.Log(LogLevel.DEBUG, "Received flush signal.");
+                        FlushQueue();
+                        continue;
+                    }
+
+                    if (item is UserEvent userEvent)
+                        AddToBatch(userEvent);
+                }
+            }
+            catch (InvalidOperationException e)
+            {
+                // An InvalidOperationException means that Take() was called on a completed collection
+                Logger.Log(LogLevel.DEBUG, "Unable to take item from eventQueue: " + e.GetAllMessages());
+            }
+            catch (Exception exception)
+            {
+                Logger.Log(LogLevel.ERROR, "Uncaught exception processing buffer. Error: " + exception.GetAllMessages());
+            }
+            finally
+            {
+                Logger.Log(LogLevel.INFO, "Exiting processing loop. Attempting to flush pending events.");
+                FlushQueue();
+            }
+        }
+
+        public void Flush()
+        {
+            EventQueue.Add(FLUSH_SIGNAL);
+        }
+
+        private void FlushQueue()
+        {
+            if (CurrentBatch.Count == 0)
+            {
+                return;
+            }
+
+            List<UserEvent> toProcessBatch = null;
+            lock (mutex)
+            {
+                toProcessBatch = new List<UserEvent>(CurrentBatch);
+                CurrentBatch.Clear();
+            }
+            
+
+            LogEvent logEvent = EventFactory.CreateLogEvent(toProcessBatch.ToArray(), Logger);
+
+            NotificationCenter?.SendNotifications(NotificationCenter.NotificationType.LogEvent, logEvent);
```

Comment:
Please split up the notifications into another PR so we can review them independently.